// proppy-react/lib/ProppySubscription.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    }
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var proppy_1 = require("proppy");
var ProppySubscription = /** @class */ (function (_super) {
    __extends(ProppySubscription, _super);
    function ProppySubscription(props) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var _this = _super.apply(this, [props].concat(args)) || this;
        var providers = props.providers, proppyFactory = props.proppyFactory, parentProps = props.parentProps;
        _this._parent = proppy_1.create({
            initialize: function () {
                this.set(parentProps);
            },
        })(providers);
        _this._p = props.proppyFactory(providers, _this._parent);
        _this.state = {
            proppyProps: _this._p.props,
        };
        return _this;
    }
    // @TODO: this needs attention
    ProppySubscription.prototype.UNSAFE_componentWillReceiveProps = function (nextProps) {
        this._parent.set(nextProps.parentProps);
    };
    ProppySubscription.prototype.componentDidMount = function () {
        var _this = this;
        this._p.subscribe(function (proppyProps) { return _this.setState({
            proppyProps: proppyProps,
        }); });
    };
    ProppySubscription.prototype.componentWillUnmount = function () {
        this._p.destroy();
    };
    ProppySubscription.prototype.render = function () {
        return this.props.children(this.state.proppyProps);
    };
    return ProppySubscription;
}(React.Component));
exports.ProppySubscription = ProppySubscription;
